# Exploring the Hidden Hot World of Long Non-coding RNAs in Thermophilic Fungus Using a Robust Computational Pipeline.

<p align="center">
  <img src="https://github.com/rogergsilva/structurally-identical-lncrnas/blob/main/images/image16.jpg" width="250" alt="Structurally identical tanscripts">
</p>

- [Abstract](#Abstract)
- [Transcriptome pipeline](#Transcriptome-pipeline)
  - [Requirements](#Requirements)
    - [Step 1: Download and install miniconda](#step-1-download-and-install-miniconda)
    - [Step 2: Setup bioconda channelss](#step-2-setup-bioconda-channels)
    - [Step 3: Installation and requirements](#step-3-installation-and-requeriments)
    - [Step 4: Create a folder for your project in your computer](#step-4-create-a-folder-for-your-project-in-your-computer)
    - [Step 5: Download the pipeline script file](#step-5-download-the-pipeline-script-file)
    - [Step 6: How to use it](#step-6-how-to-use-it)    
- [Input and Output](#Input-Output)
- [Example of how to use this pipeline](#Example-of-how-to-use-this-pipeline)
- [Bugs](#Bugs)
- [Citation](#Citation)

# Abstract

Long noncoding RNAs (lncRNAs) are versatile RNA molecules recently identified as key regulators of gene expression in response to environmental stress. Our primary focus in this study was to develop a robust computational pipeline for identifying structurally identical lncRNAs across replicates from publicly available bulk RNA-seq datasets. To demonstrate the effectiveness of the pipeline, we utilized the transcriptome of the thermophilic fungus Thermothelomyces thermophilus and assessed the expression pattern of lncRNA in conjunction with Heat Shock Proteins (HSP), a well-known family critical for the cell's response to elevated temperatures. Our results demonstrate that the identification of structurally identical transcripts among replicates in this thermophilic fungus ensures the reliability and accuracy of RNA studies, contributing to the validity of biological interpretations. Furthermore, the majority of lncRNAs exhibited a distinct expression pattern compared to HSPs. We hope this investigation contributes to advancing our comprehension of the biological mechanisms involving lncRNAs in thermophilic fungi.

# Identification of identical lncRNAs pipeline

## Requirements
* Internet access;
* Python installed (> 3.10) - https://www.python.org/

### Step 1: Download and install miniconda
Please, see https://docs.anaconda.com/free/miniconda/index.html

### Step 2: Setup bioconda channels
After having installed miniconda, you will need to set up the bioconda channels.
Follow the Usage steps on https://bioconda.github.io/

## Setp 3: Installation and requirements
### Requirements

The following python libraries must be installed on your machine:
```shell
  conda install statistics, biopython
```
### Step 4: Create a folder for your project in your computer
Create a folder in your computer where you will carry out your in-silico experiments.
For example, if your project will name PRJ123, you will execute the following command:
Linux operation system
```shell
mkdir PRJ123
```
### Step 5: Download the pipeline script file
Download the file identical.py from the Github and save it in the folder you have created. 
For example, save it in the folder PRJ123.

## Step 6: How to use it
Open a command line terminal, change the path to the folder PRJ123 and execute the following command:
```shell
cd PRJ123
python identical.py -t /home/user/ALL_merged_track.tracking
                    -f /home/user/ALL_merged_track.fasta
                    -l /home/user/lncrna_ids.txt
```

All the options
```shell
python identical.py -h
usage: identical.py [-h] [-l LNCRNA] [-t TRACKING] [-f FASTA] [-o OUTPUT] [-g GROUPS]

Identify structurally identical transcripts.

options:
  -h, --help            show this help message and exit
  -l LNCRNA, --lncRNA LNCRNA
                        File containing only lncRNA transfrag IDs
  -t TRACKING, --tracking TRACKING
                        Tracking file from StringTie2 output to be read
  -f FASTA, --fasta FASTA
                        Fasta file from StringTie2 output to be used to save identical transcripts
  -o OUTPUT, --output OUTPUT
                        Output file (default: output.csv)
  -g GROUPS, --groups GROUPS
                        Group your samples in the same order they were processed by StringTie

```
This command will create an output.csv file in the current directory.

# Input and Output
The script requires the tracking file that was generated by StringTie. Usually, this file contains the extension .tracking.
The output csv file will have the transfrag ID, type of the transfrag, exon number, FPKM, FPKM-STD, TPM, TPM-STD, coverage, coverage STD, length, and length STD.

# Example of how to use this pipeline
All files used in this example can be found in the folder test.
```shell
python identical.py -t test/ALL_merged_track.tracking
                    -f test/ALL_merged_track.fasta
                    -l test/lncrna_ids.txt
```

# Bugs
Do you have an issue? Please file it on :https://github.com/rogergsilva/structurally-identical-lncrnas/issues

# Citation
