# Exploring the Hidden Hot World of Long Non-coding RNAs in Thermophilic Fungus Using a Robust Computational Pipeline.

<p align="center">
  <img src="https://github.com/rogergsilva/structurally-identical-lncrnas/blob/main/images/image16.jpg" width="250" alt="Structurally identical tanscripts">
</p>

- [Abstract](#Abstract)
- [Transcriptome pipeline](#Transcriptome-pipeline)
  - [install with Bioconda](#install-with-bioconda)
    - [Step 1: download and build libisal](#step-1-download-and-build-libisal)
    - [step 2: download and build libdeflate](#step-2-download-and-build-libdeflate)
    - [Step 3: download and build fastp](#step-3-download-and-build-fastp)
- [Citation](#Citation)

# Abstract

Long noncoding RNAs (lncRNAs) are versatile RNA molecules recently identified as key regulators of gene expression in response to environmental stress. Our primary focus in this study was to develop a robust computational pipeline for identifying structurally identical lncRNAs across replicates from publicly available bulk RNA-seq datasets. To demonstrate the effectiveness of the pipeline, we utilized the transcriptome of the thermophilic fungus Thermothelomyces thermophilus and assessed the expression pattern of lncRNA in conjunction with Heat Shock Proteins (HSP), a well-known family critical for the cell's response to elevated temperatures. Our results demonstrate that the identification of structurally identical transcripts among replicates in this thermophilic fungus ensures the reliability and accuracy of RNA studies, contributing to the validity of biological interpretations. Furthermore, the majority of lncRNAs exhibited a distinct expression pattern compared to HSPs. We hope this investigation contributes to advancing our comprehension of the biological mechanisms involving lncRNAs in thermophilic fungi.

# Transcriptome pipeline

# Citation
